### PR TITLE
feat(otelcol/exporter/splunkhec) Add Splunk HTTP Event Collector (HEC) Exporter 

### LINKED ISSUE #1646: Splunk HEC Exporter Integration for Grafana Alloy
### Request

This document proposes the addition of a feature that enables Grafana Alloy to forward logs directly to a Splunk instance using the Splunk HTTP Event Collector (HEC) protocol. This functionality would be achieved by integrating the Splunk HEC exporter from the [OpenTelemetry Collector Contrib Repository](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/splunkhecexporter) into Grafana Alloy.

### Use case

Many organizations, including ours, leverage Splunk as their primary log management solution. Currently, Grafana Alloy lacks a native way to send collected logs directly to Splunk. This necessitates workarounds, such as deploying a separate log forwarder. These workarounds introduce additional complexity, potential points of failure, and hinder a streamlined observability experience.

### Future Considerations
- Explore potential integrations with other popular log management solutions beyond Splunk.
- Investigate the feasibility of supporting additional data formats beyond logs (e.g., metrics) through the Splunk HEC exporter.

### Conclusion
Integrating the Splunk HEC exporter into Grafana Alloy would significantly improve the platform's log forwarding capabilities. This feature would streamline data flow, enhance user experience, and improve overall observability within the Grafana ecosystem. We believe this integration aligns with Grafana's commitment to providing a comprehensive and user-friendly observability platform.

### Community component maintainer

- [x] I agree to be a maintainer for this component.

### Community component description

The proposed integration with the Splunk HEC exporter would address these challenges by providing the following benefits:

- **Simplified Log Forwarding:** Eliminates the need for a separate log forwarder, streamlining data flow and reducing operational overhead.
- **Improved Efficiency:** Enables direct communication between Grafana Alloy and Splunk, potentially enhancing data transfer speed and reducing latency.
- **Enhanced User Experience:** Provides a more unified observability experience by allowing users to manage log collection and forwarding within Grafana Alloy.

### Technical Considerations
- The integration should leverage the existing Splunk HEC exporter from the OpenTelemetry Collector Contrib Repository.
- Configuration options within Grafana Alloy should allow users to specify:
  - Splunk HEC endpoint URL
  - Splunk HEC token
  - Relevant TLS settings (if applicable)
- Error handling and logging mechanisms should be implemented to ensure proper feedback for troubleshooting purposes.

### PR DESCRIPTION
**Pull request by adlotsof**
#### PR Description

Adds functionality described in https://github.com/grafana/alloy/pull/1645 and the proposal at https://github.com/grafana/alloy/issues/1646

Pinging @PatMis16 

#### Which issue(s) this PR fixes

Fixes #1646 

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [x] Config converters updated

### CODE REVIEW

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:5]:
```suggestion
description: Learn about otelcol.exporter.splunkhec
```

New topics don't need an alias. We only need to add these if we move or rename an existing/published topic.

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:30]:
```suggestion
        token = "YOUR_SPLUNK_TOKEN"
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:52]:
```suggestion
splunk                     | [splunk][]                    | Configures the Splunk HEC exporter.                                        | yes
splunk->otel_to_hec_fields | [otel_to_hec_fields][]        | Configures mapping of Open Telemetry to HEC Fields.                        | no
splunk->telemetry          | [telemetry][]                 | Configure the exporters telemetry.                                         | no
splunk->heartbeat          | [heartbeat][]                 | Configure the exporters heartbeat settings.                                | no
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:95]:
```suggestion
`token`                    | `secret` | Splunk HEC Token.                                              |                               | yes
`log_data_enabled`         | `bool`   | Enable sending logs from the exporter.                          | `true`                        | no
`profiling_data_enabled`   | `bool`   | Enable sending profiling data from the exporter.                | `true`                        | no
`source`                   | `string` | [Splunk Source](https://docs.splunk.com/Splexicon:Source).      | `""`                          | no
`source_type`              | `string` | [Splunk Source Type}(https://docs.splunk.com/Splexicon:Sourcetype). | `""`                          | no
`index`                    | `string` | Splunk Index Name.                                               | `""`                          | no
`disable_compression`      | `bool`   | Disable GZip compression.                                        | `false`                       | no
`MaxContentLengthLogs`     | `int`    | Maximum log payload size in bytes.                               | `2097152`                     | no
`MaxContentLengthMetrics`  | `int`    | Maximum metric payload size in bytes.                            | `2097152`                     | no
`MaxContentLengthTraces`   | `int`    | Maximum trace payload size in bytes.                            | `2097152`                     | no
`MaxEventSize`             | `int`    | Maximum event payload size in bytes.                            | `5242880`                     | no
`splunk_app_name`          | `string` | Used to track telemetry for Splunk Apps by name.                | `Alloy`                       | no
`splunk_app_version`       | `string` | Used to track telemetry by App version.                          | `""`                          | no
`otel_to_hec_fields`       | `block`  | Creates a mapping from attributes to HEC fields.                 | `null`                        | no
`health_path`              | `string` | Path for the health API.                                         | `/services/collector/health'` | no
`health_check_enabled`     | `bool`   | Used to verify Splunk HEC health on exporter startup.     | `true`                        | no
`export_raw`               | `bool`   | Send only the logs body when targeting HEC raw endpoint.         | `false`                       | no
`use_multi_metrics_format` | `bool`   | Use multi-metrics format to save space during ingestion.        | `false`                       | no
`heartbeat`                | `block`  | Configuration to enable heartbeat.                               | `false`                       | no
`telemetry`                | `block`  | Configuration for Splunk HEC exporter telemetry.                  | `null`                        | no
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:102]:
```suggestion
`severity_text`            | `string` | Maps severity text field to a specific HEC field.                |   `""`                        | no
`severity_number`          | `string` | Maps severity number field to a specific HEC field.              |   `""`                        | no
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:110]:
```suggestion
`interval`                 | `int`    | Time interval for the heartbeat interval, in seconds.            |   `0`                         | no
`startup`                  | `bool`   | Send heartbeat events on exporter startup.                       |   `false`                        | no
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:121]:
```suggestion
`enabled`                  | `bool`                | Enable telemetry inside the exporter.                            |   `false`                     | no
`override_metrics_names`   | `map[string]string`   | Override metrics for internal metrics in the exporter.           |   `map[string]{}`             | no

```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:130]:
```suggestion
`endpoint`                | `string`   | The Splunk HEC endpoint to use.                                              |         | yes
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:170]:
```suggestion
`otelcol.exporter.splunkhec` is only reported as unhealthy if given an invalid configuration.
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:175]:
```suggestion
`otelcol.exporter.splunkhec` does not expose any component-specific debug information.
```

Comment (adlotsof) [reply docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md]: Thanks a lot for the detailed review!

**Comment by adlotsof** [comment]
I assume the failed ci is due to me forgetting to re-run `go generate ./internal/tools/docs_generator/` after updating the docs?

File: internal/component/otelcol/exporter/splunkhec/splunkhec.go
Comment (wildum) [inline internal/component/otelcol/exporter/splunkhec/splunkhec.go:2]: if you volunteer to maintain the component, you should put your @ here like in the datadog community component:
// Maintainers for the Grafana Alloy wrapper:
//	- @adlotsof

File: internal/component/otelcol/exporter/splunkhec/config/splunkhec.go
Comment (wildum) [inline internal/component/otelcol/exporter/splunkhec/config/splunkhec.go:109]: did you test the heartbeat functionality? is it working as expected?

File: internal/component/otelcol/exporter/splunkhec/splunkhec.go
Comment (wildum) [inline internal/component/otelcol/exporter/splunkhec/splunkhec.go:64]: in the validate we should have the same checks as in the otel config. I found out that the config is invalid if LogDataEnabled and ProfilingDataEnabled are both set to false. This should be documented in the .md file

File: internal/component/otelcol/exporter/splunkhec/config/splunkhec.go
Comment (wildum) [inline internal/component/otelcol/exporter/splunkhec/config/splunkhec.go:40]: this won't work, we need an Alloy struct for the batcher config

File: internal/component/otelcol/exporter/splunkhec/splunkhec.go
Comment (adlotsof) [reply internal/component/otelcol/exporter/splunkhec/splunkhec.go:2]: Realistically, i will have to discuss at work (with @PatMis16) - I would love to have the time in my free time, but its just not the case at the moment. I will come back to you on the `volunteering to be maintainer` - question.

**Comment by adlotsof** [comment]
> Thanks for your contributions, it looks great already :) 

Puh i pretty much just copied (or tried to copy) what was done in the [datadog PR](https://github.com/grafana/alloy/pull/739) . I will address your comments as soon as i manage to also work on them. Thanks a lot for the kind words!

> The config converter is missing (this is used to convert an otel config to an alloy config automatically). Here is the converter folder: https://github.com/grafana/alloy/tree/main/internal/converter/internal/otelcolconvert and here is a PR that adds a new converter: https://github.com/grafana/alloy/pull/739/files 

I dont know how i missed this - but i will add the converter when addressing the remaining, open comments.

> Else, are you ok to become a maintainer of this component?

I will come back to you on this

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:123]:
```suggestion
`max_size_items`   | `uint`   | Maximum number of batch items, if the batch exceeds this value, it will be broken up into smaller batches. Must be greater than or equal to min_size_items. Setting this value to zero disables the maximum size limit.          |  `0`           |  no
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:50]: Looks like this is missing the cell in the last column for required yes/no

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:77]:
```suggestion
`log_data_enabled`           | `bool`   | Enable sending logs from the exporter. One of `log_data_enabled` or `profiling_data_enabled` must be `true`.                         | `true`                        | no
`profiling_data_enabled`     | `bool`   | Enable sending profiling data from the exporter. One of `log_data_enabled` or `profiling_data_enabled` must be `true`.             | `true`                        | no
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:80]:
```suggestion
`source`                     | `string` | [Splunk source](https://docs.splunk.com/Splexicon:Source).      | `""`                          | no
`source_type`                | `string` | [Splunk source sype}(https://docs.splunk.com/Splexicon:Sourcetype). | `""`                          | no
`index`                      | `string` | Splunk index name.                                               | `""`                          | no
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:85]:
```suggestion
`max_content_length_logs`    | `uint`   | Maximum log payload size in bytes. Must be less than 838860800 (~800MB).                               | `2097152`                     | no
`max_content_length_metrics` | `uint`   | Maximum metric payload size in bytes. Must be less than 838860800 (~800MB).                           | `2097152`                     | no
`max_content_length_traces`  | `uint`   | Maximum trace payload size in bytes. Must be less than 838860800 (~800MB).                           | `2097152`                     | no
`max_event_size`             | `uint`   | Maximum event payload size in bytes. Must be less than 838860800 (~800MB).                            | `5242880`                     | no
```

File: docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md
Comment (clayton-cornell) [inline docs/sources/reference/components/otelcol/otelcol.exporter.splunkhec.md:122]:
```suggestion
`flush_timeout`             | `time.Duration`       | The time after which a batch will be sent regardless of its size.    | `200ms`  | no
`min_size_items`   | `uint`   |  The number of items at which the batch is sent regardless of the timeout.          |  `8192`              | no
```

**Review by clayton-cornell — COMMENTED** [review]
Some additional doc suggestions

File: internal/component/otelcol/exporter/splunkhec/config/splunkhec.go
Comment (wildum) [inline internal/component/otelcol/exporter/splunkhec/config/splunkhec.go:250]: nit: I find `if !args.LogDataEnabled && !args.ProfilingDataEnabled` easier to read but feel free to keep it like this if you prefer

**Review by wildum — COMMENTED** [review]
Awesome work! I resolved the comments that were addressed, there are only a few left but we are very close to get it merged.
We still need to find a maintainer though

**Review by wildum — APPROVED** [review]

**Comment by wildum** [comment]
All good from my POV, please ignore the flaky test. One last thing before we merge: I see that you added yourself and @PatMis16 as maintainers. We update our Otel dependencies ~every 6 weeks (before every release). A few days before we cut the release candidate, we will ping you on the community slack and on github to give you the opportunity to sync the hec exporter with upstream.
But no pressure:
- it's **okay** to miss some updates
- it's **okay** if the component does not have all upstream's features
- you can step out at any time from being a maintainer
If there are breaking changes in the component that are blocking our release and that no maintainers are available we can step in and help. If there is no easy way for us to solve them we may temporarily disable the component.
The way we update the otel dependencies is not ideal but it's something that we are working on improving.

@adlotsof @PatMis16 if you are ok with this, please thumbs up this message and I will proceed to merge the PR.

If you need to reach out to the team you can message us on slack directly or ping us in the #alloy slack channel

**Comment by PatMis16** [comment]
Guess it's ok.

**Pull request merged**
